frontend/catalyst/jax_primitives.py

```diff
@@ -1195,26 +1224,41 @@ def emit_branches(preds, branch_jaxprs, ip):
 
 #
 # while loop
-#
-@while_p.def_abstract_eval
-def _while_loop_abstract_eval(*args, cond_jaxpr, body_jaxpr, **kwargs):
-    return body_jaxpr.out_avals
-
-
+# pylint: disable=too-many-arguments
 @while_p.def_impl
 def _while_loop_def_impl(
-    ctx, *iter_args_plus_consts, cond_jaxpr, body_jaxpr, cond_nconsts, body_nconsts
+    ctx,
+    *iter_args_plus_consts,
+    cond_jaxpr,
+    body_jaxpr,
+    cond_nconsts,
+    body_nconsts,
+    nimplicit,
+    preserve_dimensions,
 ):  # pragma: no cover
     raise NotImplementedError()
 
 
+@while_p.def_abstract_eval
+def _while_loop_abstract_eval(*in_type, body_jaxpr, nimplicit, preserve_dimensions, **kwargs):
+    _assert_jaxpr_without_constants(body_jaxpr)
```

Comment:
What is the implication of this restriction? Do you have an example?

Reply:
For me this is more a self-documenting code. Literally, body jaxpr of a loop should not have any constants, all constants should be moved to arguments using `convert_constvars_jaxpr`, so we do it in our new generic tracing function [here](https://github.com/PennyLaneAI/catalyst/blob/f03253d8e5ffeee5b91ba79514057d8d813bc874/frontend/catalyst/utils/jax_extras.py#L385)

Reply:
Sure the function name makes it very clear what is being asserted, what is not clear is why. Did we assert this previously? If not, what changed? What is the resulting effect of asserting this?

This is what I wanted to find out with my comment.

Reply:
We did not assert this previously, nor did we rely on this fact here. With dynamic dimensions, we need to know the number of constants to process type links correctly. So I suggest to add this assertion of having no constants at this point.

Reply:
Am I understanding you correctly that we are introducing a regression that prevents a user from accessing variables inside the body of while loop that are not passed in via the arguments? For example, this would no longer be allowed:

```py
x = 5

@while_loop(lambda ...)
def f(y):
  new_y = x * y
  return new_y
```

Reply:
I think it would be correct to say that we are not going to use arrays with dynamic shapes in this mode. Operations on the statically-shaped arrays should be supported, so I thought about it as of limitation, not a regression.

Reply:
So you are saying that variable capture (which jax calls constants) is only prevented when the dynamic array feature is active? 

Unless I'm missing something, this assertion is applied to all while loop primitives.

Reply:
> So you are saying that variable capture (which jax calls constants) is only prevented when the dynamic array feature is active?
Yes, I think so. Look [here](https://github.com/PennyLaneAI/catalyst/blob/2cf433c2aa009feccb7e63213ffb8bfaaf541558/frontend/catalyst/utils/jax_extras.py#L385) - we move constants to variables at the time of `out_sig.out_jaxpr()` [here](https://github.com/PennyLaneAI/catalyst/blob/2cf433c2aa009feccb7e63213ffb8bfaaf541558/frontend/catalyst/pennylane_extensions.py#L1953)

Reply:
Looks like you are right, tests with outer variable capture seem to run just fine.

In that case I don't understand what input would trip up the assertion. Do you have an example?

Reply:
A failing example would be a `body_jaxpr` for which someone forgot to convert constants into inputs. Moving constants into inputs means that one takes the responsibility for remembering the constant values somewhere outside the Jaxpr data. We need to do this because constants need to be lowered too. I think it is correct to say that this assertion checks that we indeed do this.

Reply:
Ok so I tried digging into this concept of jaxpr constants (`constvars`) to figure out when they are actually used. The page on "understanding jaxpr" was helpful in this regard: https://jax.readthedocs.io/en/latest/jaxpr.html#constant-vars
Some insights:
- `constvars` are only used for constant (i.e. non-tracer) _array_ values, that is scalars are embedded directly into the jaxpr
- `constvars` are used for values that appear inside a function which are _hoisted out_ as jaxpr arguments, outside variable capture has nothing to with it
- `constvars` are hoisted out all the way to the top-level jaxpr, so they shouldn't appear in higher order primitives (primitives with their own jaxpr)

Does this match your understanding? Did you find, manipulate, or otherwise interact with jaxpr constants while working on this feature? From my testing I was only ever able to generate constants in the top-level jaxpr anyways (so the jaxpr for the qjit function, never in primitive/nested jaxprs).

Reply:
> I would rather say that constants are not used for literal scalars, but might be used for other values. The to_jaxpr2 tracer does return outer tracers as constants of the inner (body) program. See the [related function](https://github.com/google/jax/blob/35071c74be259b8018b46ed4aafd624b179586e4/jax/_src/interpreters/partial_eval.py#L1808)

That's what I said no? (constant) scalar values never end up as `constvars`, because they are embedded directly into the IR as literals. 

> I believe that we do not use this mechanism for our primitives. Compare with the code for Jax' own primitives [here](https://github.com/google/jax/blob/35071c74be259b8018b46ed4aafd624b179586e4/jax/_src/lax/lax.py#L2422) where they do use some constants folding machinery.

The discussion is about constant hoisting, I don't see how folding is related. 

> I think that Jax does not treat our primitives as higher order ones. This is because we pass jaxprs as keyword args without letting Jax know about it in any way. So, I believe that all our body jaxprs are top-level in this sense, and we are responsible for dealing with constants at every layer.

I have seen no evidence of this. I have tested examples where variables are hoisted out to the top-level scope when using our primitives.

Reply:
> That's what I said no? (constant) scalar values never end up as constvars, because they are embedded directly into the IR as literals.

Hmm, I think you said "constvars are only used for constant (i.e. non-tracer) array values" while I am saying that constants can be used for tracers. And it is exactly the case where we capture tensor variables from the outside scope. 

>  I have tested examples where variables are hoisted out to the top-level scope when using our primitives.

Can you share some examples where constants propagated through more than one level of Jaxpr? (I am worrying about possible misunderstanding as well)

Reply:
> Hmm, I think you said "constvars are only used for constant (i.e. non-tracer) array values" while I am saying that constants can be used for tracers. And it is exactly the case where we capture tensor variables from the outside scope.

Thanks for clarifying! Look at the description in the [jax docs](https://jax.readthedocs.io/en/latest/jaxpr.html#constant-vars), I'm pretty sure `constvars` are only ever used for constant values in the program (not values that where originally tracers).

> And it is exactly the case where we capture tensor variables from the outside scope.

This was the second insight above, `constvars` has nothing to do with capturing variables from the outside scope (which I originally believed). Instead it is used in connection with hoisting variables up into an outer scope. I think this is reasonable well explained in the [jaxpr doc page](https://jax.readthedocs.io/en/latest/jaxpr.html#understanding-jaxprs), but you have to look for mentions of `constvars`.

> Can you share some examples where constants are propagated through more than one level of Jaxpr? (I am worrying about possible misunderstanding as well)

For example: (you can add more nesting with same effect)

```py
@qjit
def f(x: int):
    @for_loop(0, 5, 1)
    def g(i):
        y = jnp.array([0.0, 1.0])
        debug.print(y)
    return g()
f.jaxpr
```

In the jaxpr you can see the float array is hoisted out to `a:f64[2]` before the `;`, so this argument is stored in the `consts` field of the outermost jaxpr:
```
{ lambda a:f64[2]; b:i64[]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; c:f64[2] d:i64[]. let
           = debug_print[memref=False] c
        in () }
      body_nconsts=1
    ] 0 5 1 a 0
  in () }
```

If `y = jnp.array(0.1)` is a scalar on the other hand, it is stored as a literal:
```
{ lambda ; a:i64[]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; b:i64[]. let
           = debug_print[memref=False] 0.1
        in () }
      body_nconsts=0
    ] 0 5 1 0
  in () }
```

And if `y = i` is a dynamic value, then it just uses that jaxpr variable (wherever it's from):
```
{ lambda ; a:i64[]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; b:i64[]. let
            = debug_print[memref=False] b
        in () }
      body_nconsts=0
    ] 0 5 1 0
  in () }
```

Reply:
Thanks @dime10. Consider this example:

``` python
@qjit
def f(x: int, y:jax.core.ShapedArray([2], int)):
    @for_loop(0, 5, 1)
    def g(i):
        debug.print(y)
    return g()

f.jaxpr
```

``` result
{ lambda ; a:i64[] b:i64[2]. let
     = for_loop[
      apply_reverse_transform=False
      body_jaxpr={ lambda ; c:i64[2] d:i64[]. let
           = debug_print[memref=False] c
        in () }
      body_nconsts=1  # <--- [1]
    ] 0 5 1 b 0
  in () }
```
 
The `y` here is a regular argument, but note the Jaxpr line marked with `[1]`: The `body_ncosts` equal to one means that Jaxpr tracer originally created a constant that was moved to argument by us just before the primitive lowering. At the time of the tracing, this constant stored a tracer. (Note, that I'm running Catalyst 0b1335c428f1, i.e. I am not in this PR's branch).

Reply:
Yes, it is our variable indeed. But it has value of one which means that 1) Jax tracer emitter body jaxpr with one Jax constvar 2) We converted it into a regular parameter and recorded this fact by increasing the variable.

When I wrote about variable capturing, I meant that if we refer outside-scope variables in a Python program, Jax emits Jaxpr program with constants. Additionally, Jax creates a datatype called ClosedJaxpr, containing the values for these constants, in this case - the outside-scope tracers. But this is not compatible with our StableHLO lowerer, which requires tracers to be passed as positional arguments only. So we convert ClosedJaxpr into a ClosedJaxpr-with-empty-constansts by calling the function named `convert_constvars_jaxpr`. The assertion you mentioned just checks that this conversion really happened - it is something that Jax does not do by default.